Add mix debug mode. Close #4106.

tuvistavie opened the pull request:
I implemented mix debug option as described in #4106.

tuvistavie commented on lib/mix/lib/mix/cli.ex:
```diff
@@ -9,6 +9,7 @@ defmodule Mix.CLI do
     Mix.Local.append_paths
 
     if System.get_env("MIX_QUIET"), do: Mix.shell(Mix.Shell.Quiet)
+    if System.get_env("MIX_DEBUG"), do: Mix.debug(true)
```
Where should I document this? I found that `MIX_QUIET` is described in the man page, 
so it would be the right place to add info about this right?

josevalim replied:
In the man page at least. We could probably add an environment variables section to the `Mix` module docs too. Could you please act on both? :)

tuvistavie commented:
I would like to discuss a point. `MIX_QUIET` and `MIX_DEBUG` now only check if the variable is set,
which means that `MIX_DEBUG=0` and `MIX_DEBUG=1` behaves exactly the same way.
I find this behavior a little counter-intuitive, so maybe having a list of blacklist values, (no, false, 0 would be a good start)
that would avoid settings the flag even if the env variable is on would be good to have.
Any thoughts?

josevalim commented:
> that would avoid setting the flag even if the env variable is on would be good to have.

Let's check if it is "1" or "true".

josevalim commented on lib/mix/lib/mix/task.ex:
```diff
@@ -275,6 +277,12 @@ defmodule Mix.Task do
   end
 
+  defp output_task_debug_info(task, args) do
+    Mix.shell.info("Running 'mix #{task_to_string(task, args)}'")
```
Can we do: `"** Running mix #{task_to_string(task, args)} (inside #{inspect proj})"`. I don t think we need to wrap in quotes the mix command, it would be just noise in this case. :)

josevalim replied:
Btw, the `proj` may be nil, so we need to check accordingly. :)

josevalim commented on lib/mix/test/mix/task_test.exs:
```diff
@@ -26,6 +26,17 @@ defmodule Mix.TaskTest do
+      "Running 'mix hello'"
+
+    Mix.shell(Mix.Shell.Process)
+    Mix.debug(false)
```
Move those two last commands to an after clause:

```
after
  Mix.shell(Mix.Shell.Process)
  Mix.debug(false)
end
```

josevalim commented:
Thank you! I have added some comments! :heart:

tuvistavie commented:
@josevalim Thanks for the feedback!
i just updated my PR.
I split the doc update in `Mix` module to another commit as it was not specific to the `DEBUG` flag.

lexmag commented on lib/mix/lib/mix.ex:
```diff
@@ -235,6 +252,20 @@ defmodule Mix do
+  @doc """
+  Sets Mix debug mode.
+  """
+  def debug(debug) do
```
I think it worth to have `is_boolean/1` guard.

tuvistavie replied:
It does seem safer! I'll add this, thank you.

lexmag commented on lib/mix/lib/mix/task.ex:
```diff
@@ -275,6 +277,13 @@ defmodule Mix.Task do
   end
 
+  defp output_task_debug_info(task, args, proj) do
+    proj_str = if is_nil(proj), do: "", else: " (inside #{inspect proj})"
```
We could make an opposite order of `if` clauses:
```elixir
if proj, do: " (inside #{inspect proj})", else: ""
```

tuvistavie replied:
I usually stick with conditions that return booleans when possible but I agree

```elixir
if proj, do: " (inside #{inspect proj})", else: ""
```

read better.

lexmag commented on lib/mix/test/mix/cli_test.exs:
```diff
@@ -52,6 +52,15 @@ defmodule Mix.CLITest do
       assert contents =~ "Hello from MyProject!\n"
       refute contents =~ "This won't appear"
+
+      contents = mix ~w[my_hello], [{"MIX_QUIET", "0"}]
```
Very nitpicking: can we be consistent in `w` sigil definition and use `()`?

tuvistavie replied:
All the sigil definitions in this file were using `~w[]` so I preferred to stick with it.
After checking, `cli_test.exs` and `test_helpers.exs` are the only files which use
`~w[]`, so we could fix this, but I think it would be better to make it as another PR,
as it is not really related to the original issue anymore.

lexmag replied:
Totally agree. :+1:

josevalim commented on lib/mix/lib/mix.ex:
```diff
@@ -152,6 +152,23 @@ defmodule Mix do
 
   Note aliases do not show up on `mix help`.
+
+  ## Environment variables
```
Sorry for the nitpick but please add new line after tiles. :)

josevalim commented on lib/mix/lib/mix.ex:
```diff
@@ -152,6 +152,23 @@ defmodule Mix do
+
+  Mix responds to the following variables:
+
+  * `MIX_ARCHIVE` - allows specifying the directory into which the archives should be installed
```
Indent bullets with 2 spaces.

josevalim commented on lib/mix/lib/mix/cli.ex:
```diff
@@ -71,6 +72,10 @@ defmodule Mix.CLI do
     end
   end
 
+  defp env_variable_activated?(name) do
```
:heart:

josevalim commented on lib/mix/lib/mix/task.ex:
```diff
@@ -275,6 +277,13 @@ defmodule Mix.Task do
 
+  defp output_task_debug_info(task, args, proj) do
+    proj_str = if proj, do: " (inside #{inspect proj})", else: ""
+    Mix.shell.info("** Running mix #{task_to_string(task, args)}" <> proj_str)
```
What about:

    "** Running mix " <> task_to_string(task, args) <> project_to_string(prog)

lexmag replied:
Then we probably don't need `output_task_debug_info` function.

tuvistavie replied:
I moved `project_to_string` to its own function, but I would rather keep 
`output_task_debug_info` as a function to reduce the noise to `run_task/3`.

josevalim commented:
@tuvistavie this is perfect. I have added one last batch of nitpicking commands and we are good to go. :)

tuvistavie commented:
@josevalim Thank you, I just pushed the last fixes.

lexmag commented on lib/mix/lib/mix/task.ex:
```diff
@@ -275,6 +277,14 @@ defmodule Mix.Task do
+  defp output_task_debug_info(task, args, proj) do
+    Mix.shell.info("** Running mix " <> task_to_string(task, args) <> project_to_string(proj))
+  end
+  defp project_to_string(nil), do: ""
```
Very last comment, could you separate function definition with newlines. :bowtie:

tuvistavie replied:
Done! :+1:

lexmag replied:
:yellow_heart:

josevalim commented:
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:
